📖 Cluster API Add-on Orchestration proposal (#6905)
kubernetes-sigs/cluster-api: docs/proposals/20220712-cluster-api-addon-orchestration.md

While building this solution, we foresee at least three iterations:

Foundation (the one described in this document), with the goal to provide a minimal viable solution for this space, integration with the package management tool of choice - Helm will be used as example -, exploring and addressing basic concern of orchestration like e.g. identifying a single source of truth for add-configuration and reconciliation mechanism to ensure the expected configuration is applied;

Advanced orchestration. As soon as the foundation for addon orchestration will be in place, we will start exploring and addressing advanced orchestration scenarios, like e.g. managing add-on to be updated before the cluster upgrade starts; This work will leverage the lifecycle hooks introduced in Cluster API v1.2.

Add-on dependencies. As a final step of this work, we are going to explore and manage use cases where there are dependencies between add-ons, and eventually introduce capabilities to coordinate add-on management accordingly (e.g upgrade addons in a certain order).

> [^1]: Package management tools usually have broader scope that goes over and behind the act of installing a Kuberentes application, like e.g. defining formats for packaging Kubernetes applications, defining repository/marketplaces where users can find libraries of existing applications, templating systems for customizing applications etc.

Reviewer: formatting looks weird, should this be at the bottom of the doc or in the previous paragraph?

- There should be support for out-of band upgrades, e.g. for CVE fix (not linked to the cluster lifecycle).
- It should be possible to orchestrate pre/post installation steps for each addon, thus allowing to manage migrations/operations like:
  - CPI migrating from in-tree to out-of-tree
  - CoreDNS configuration migration from one version to another
- Addon deletion should be orchestrated before Cluster deletion

The current proposal starts to address part of this problem by designing how Cluster API can orchestrate an external tool for addon management; the external tool is expected to provide all the foundational capabilities for addon management like e.g. addon packages/repository, templating/configuration, management of the creation, upgrade and deletion workflow etc.

Cluster API - and most specifically a new Cluster API-aware component described in this proposal - should orchestrate the above tool so that Cluster lifecycle and addon lifecycle are consistent.

### Goals

> - To design a solution for orchestrating Cluster add-on

Reviewer: ```suggestion
- To design a solution for orchestrating Cluster add-ons
```
?

> - To design a solution for orchestrating Cluster add-on

Reviewer: ```suggestion
- To design a solution for orchestrating Cluster add-ons
```
?

> - To leverage an existing package management tool, Helm in this case, for all the foundational capabilities of addon management like e.g. addon packages/repository, templating/configuration, addon creation, upgrade and deletion etc. 

Reviewer: ```suggestion
- To leverage existing package management tools, for all the foundational capabilities of addon management like e.g. addon packages/repository, templating/configuration, addon creation, upgrade and deletion etc. 
```

Reviewer: - To make addon management in CAPI modular and pluggable and to make it easy for anyone to build a CAPI addon provider based on their favorite package management tool, just like infrastructure and bootstrap providers.

### Non goals

> - To implement a full fledged package management tool in Cluster API; there are already several awesome package management tools in the ecosystem, and CAPI should not reinvent the wheel. 

Reviewer: - To force users to use one particular package management tool.

> - To provide mechanism for altering, customizing or dealing with single Kubernetes resources defining a Cluster add-on (Deployments, Services, ServiceAccounts etc); CAPI should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool.

Reviewer: ```suggestion
- To provide a mechanism for altering, customizing, or dealing with single Kubernetes resources defining a Cluster add-on (Deployments, Services, ServiceAccounts, etc); CAPI should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool.
```

### Future work

- Provide support for advanced orchestration use cases leveraging on Cluster API’s lifecycle hooks for Cluster API.

> - Introduce capabilities to coordinate operation accordingly to Add-on dependencies.

Reviewer: ```suggestion
- Introduce capabilities to reconcile addons in an order that respects dependencies between addons.
```

> - To eventually reconsider the future of ClusterResourceSet experimental features. 

Reviewer: ```suggestion
- Deprecate or remove the ClusterResourceSet experimental feature. 
```

## Proposal

> This document introduces the concept of a Cluster API addon provider using [Helm](https://helm.sh/), a new, Kubernetes CRD called HelmChartProxy to be deployed on the Management Cluster. HelmChartProxy will act as an intermediary (proxy) between Cluster API and the Helm to install packages onto workload clusters.

Reviewer: Consider making it more clear that this is just a reference implementation and not _the_ proposed solution

> ### User Stories

Reviewer: This user stories should probably not be Helm specific and should cover all addon providers

However, considering the fact that the link between Cluster lifecycle and Cluster add-on lifecycle is simpler than e.g the link between a Cluster and its infrastructure provider, in this case it is possible to adopt a simplified interaction model between Cluster API and HelmChartProxy (vs an integration model based on CRD with a set of well known field defined in the CAPI contract).

In the first iteration, we will implement this as a single, concrete HelmChartProxy CRD, with a controller watching Cluster API resources, mainly the Cluster and eventually the ControlPlane object.

This simple approach makes it possible to implement HelmChartProxy immediately and explore some foundational problems, however there will be a follow up iteration where HelmChartProxy will take advantage of Cluster API Runtime Extension and Lifecycle Hooks for supporting advanced orchestration use cases. 

Following paragraphs provide a deep dive on the responsibilities of the ClusterAddonProvider component; we are using Helm as a first target package management, but similar solutions can be defined for other tools as well. 

> ClusterAddonProvider for Helm

Reviewer: formatting?

Helm is a CNCF graduated project implementing a package manager widely used in the community; Helm packages are named charts.

ClusterAddonProvider for Helm could be implemented as a kubebuilder controller with a single reconciler watching for the Cluster object; each reconciliation will trigger Helm operations by leveraging on  https://github.com/helm/helm imported as a library.

The ClusterAddonProvider for Helm is going to implement two new CRDs (see next paragraphs for the full spec). The first is named HelmChartProxy used to specify a Helm chart with configuration values and a selector identifying the clusters the add-on applies to. The second is called HelmReleaseProxy and contains information about the underlying Helm release installed on a cluster from a Helm chart. All the configurations will be done from the HelmChartProxy and the HelmReleaseProxy is simply used to maintain an inventory as well as surface information.

> A demo of the prototype can be found in the Cluster API Office Hours recording. The prototype implements HelmChartProxy and HelmReleaseProxy as well as controllers for both CRDs.

Reviewer: link to the demo?

    kind: Cluster
    name: default-15210
    namespace: default
  ready: true
```

#### HelmReleaseProxy CRD

The second CRD, HelmReleaseProxy, is responsible for installing and orchestrating the Helm release on a selected workload Cluster. The HelmReleaseProxy is created as part of the parent HelmChartProxy flow. The parent HelmChartProxy will be responsible for keeping its HelmReleaseProxy children up-to-date as its lifecycle is updated (for example, if the `values` properties of a HelmChartProxy are updated, its child HelmReleaseProxy resources will be updated as well).

##### Spec

- `clusterRef`: An object reference to the Cluster the Helm release is installed on.
- `repoURL`: The URL of the Helm chart repository.
- `chartName`: The name of the Helm chart.
- `releaseName`: The name of the Helm release.
- `version`: The version of the Helm chart.
- `values`: A map of key/value pairs specifying values to be used when installing the Helm release. Note that this map is populated with the resolved values from the Cluster.

##### Status

- `ready`: A boolean that indicates when the Helm release is successfully installed and up to date with the HelmReleaseProxy spec.

> - `failureReason`: A string pointer indicating the error message, if applicable.

Reviewer: let's consider adding Conditions to better surface granular status details. FailureReason is technically deprecated and should theoretically be removed from other CAPI CRDs eventually (https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200506-conditions.md)

Author: Yep, adding conditions is on my TODO list, though I didn't realize FailureReason is deprecated. Is FailureMessage in the same position?

Reviewer 2: Yes

```yaml
apiVersion: addons.cluster.x-k8s.io/v1beta1
kind: HelmReleaseProxy
metadata:
  labels:
    addons.cluster.x-k8s.io/helmchartproxy-name: cloud-provider-azure-chart
    cluster.x-k8s.io/cluster-name: default-15210
  name: cloud-provider-azure-default-15210
  namespace: default
  ownerReferences:
  - apiVersion: addons.cluster.x-k8s.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: HelmChartProxy
    name: cloud-provider-azure-chart
    uid: <some-uid>
spec:

>   chartName: cloud-provider-azure

Reviewer: consider adding another addon that isn't a cloud provider as an example for illustration purposes so people can picture different use cases while reviewing this

Author: Sounds good, I'll work on putting together a different example and adding it back.

  clusterRef:
    apiVersion: cluster.x-k8s.io/v1beta1
    kind: Cluster
    name: default-15210
    namespace: default
  releaseName: cloud-provider-azure
  repoURL: https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo
  values:
    cloudControllerManager.containerResourceManagement.limitsCPU: "11"
    infra.clusterName: default-15210
  version: 1.23.8
status:
  ready: true
```

#### Controller Design

> The HelmChartProxy controller is responsible for maintaining our list of HelmReleaseProxies and resolving any referenced Cluster fields. The controller will

Reviewer: incomplete sentence or unformatted list

- To design a solution for orchestrating Cluster add-ons.
- To leverage existing package management tools such as Helm for all the foundational capabilities of add-on management, i.e. add-on packages/repository, templating/configuration, add-on creation, upgrade and deletion etc. 
- To make add-on management in Cluster API modular and pluggable, and to make it simple for developers to build a Cluster API Add-on Provider based on any package management tool, just like with infrastructure and bootstrap providers.

### Non goals

- To implement a full fledged package management tool in Cluster API; there are already several awesome package management tools in the ecosystem, and CAPI should not reinvent the wheel. 
- To provide a mechanism for altering, customizing, or dealing with single Kubernetes resources defining a Cluster add-on, i.e. Deployments, Services, ServiceAccounts. Cluster API should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool.
- To expect users to use a specific package management tool.
- To implement a solution for installing add-ons on the management cluster itself.

### Future work

> - Handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version.

Reviewer 3: > when the workload Cluster is upgraded to a new Kubernetes version.

As of v1beta1, there is no "Kubernetes version" of a Cluster. There is only the Kubernetes version of the control plane, and of groups of machines (MachineDeployment/Pool).

ClusterClass adds the notion of a Kubernetes version of an entire cluster.

Lifecycle hooks also require ClusterClass.

Is this future work limited to ClusterClass clusters?

Reviewer 4: What @Jont828 said is correct. This work will probably rely on the lifecycle hook work (and thus CC cluster only), but at this stage details are still TBD.

```go
// HelmChartProxySpec defines the desired state of HelmChartProxy.
type HelmChartProxySpec struct {
	// ClusterSelector selects Clusters with a label that matches the specified key/value pair. The Helm chart will be
	// installed on all selected Clusters. If a Cluster is no longer selected, the Helm release will be uninstalled.
	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`

	// ChartName is the name of the Helm chart in the repository.
	ChartName string `json:"chartName,omitempty"`

	// RepoURL is the URL of the Helm chart repository.
	RepoURL string `json:"repoURL,omitempty"`

	// ReleaseName is the release name of the installed Helm chart. If it is not specified, a name will be generated.
	// +optional
	ReleaseName string `json:"releaseName,omitempty"`

	// Version is the version of the Helm chart. If it is not specified, the latest version will be used.
	// +optional
	Version string `json:"version,omitempty"`

	// Namespace is the namespace the Helm release will be installed on each selected Cluster. If it is not specified, the default namespace.
	// +optional

> 	Namespace string `json:"namespace,omitempty"`

Reviewer 3: I recommend we rename this to `ReleaseNamespace` for two reasons: First, "namespace" alone does not make it clear that this is the release namespace; second, it would be consistent with `ReleaseName`, above.

Additionally, other package managers like kapp require in-cluster components such as the kapp-controller. In this case, ClusterAddonProvider should include those in-cluster components as well.

#### Mapping HelmChartProxies to Helm charts

A Helm chart is uniquely identified using the URL of its repository, the name of the chart, and the chart version. As a result, each HelmChartProxy uses the `repoURL`, `chartName`, and `version` fields to uniquely identify a Helm chart and determine which version to install. This allows HelmChartProxy to be compatible with any valid Helm repository and chart.

#### Mapping Kubernetes version to add-ons version

Cluster add-ons by definition have a lifecycle strictly linked to the Cluster lifecycle, and the main consequences of this is the fact that there is a strict correlation between the Cluster’s Kubernetes version and the add-on version to be installed in a Cluster.

> In some cases this correlation is explicit, like e.g. for CPI following the “Versioning Policy for External Cloud Providers”; in other cases the compatibility matrix is defined case by case in documentation, like for “Supported version” in the Calico documentation.

Author: @fabriziopandini By the way, what is CPI?

Reviewer 2: cloud provider interface. Not sure about the exact definitions. My assumption is roughly CPI is the interface while the components usually implementing the interface are called "cloud controller manager"

> On top of that, in many Cluster API installations, the add-on version an operator would like to use is limited to a list of certified versions defined in each organization/product vendor.

Author: @fabriziopandini Just to clarify, does this mean that add-on versions are restricted to the versions that are compatible with their existing Cluster components? Or are there additional cases we want to consider too?

Reviewer 4: AFAIK product vendors usually certify only a small combinations of component versions.
But I don't think that this requires additional work from our side at this stage

  - Provide a modular design such that the HelmChartProxy controller does not need to implement any Helm operations and only make Kubernetes client calls to create, update, or delete a HelmReleaseProxy.

## Cluster API enhancements

Despite the fact that most of the heavy lifting for add-on management is delegated to external ClusterAddonProvider components, over time we would like to provide Cluster API’s users an integrated and seamless user experience for add-on orchestration.

### clusterctl

ClusterAddonProvider is a new component to be installed in a Management Cluster, and in order to make this operation simpler, we can extend clusterctl in order to manage ClusterAddonProvider implementations in clusterctl init, clusterctl upgrade, etc.

Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.

## Alternatives

### Why not ClusterResourceSets?

This proposal introduces ClusterAddonProviders as opposed to overhauling ClusterResourceSets for the following reasons:

> - ClusterResourceSets violate the Cluster API design philosophy of defining a desired state through a declarative spec and having controllers create, delete, or update resources to build the desired state. Instead, ClusterrResourceSets are designed to "apply once and forget."

Reviewer 3: I disagree.

Community members [proposed](https://github.com/kubernetes-sigs/cluster-api/pull/6188) to extend CRS with an "apply always" strategy, i.e. to update when the desired resources have changed. This proposal is implemented in a Cluster API fork, and used in the production today.

Author: Thanks for sharing that proposal, I wasn't aware of that. Is it similar to the idea of #6555 to add a Reconcile() mode? I suppose we could afford to reword this section a bit. The main point I was trying to get across is that in the main branch of CAPI, CRS is a bit of an outlier in that it doesn't yet have a canonical Reconcile() mode and I don't believe there's a Delete() mode implemented either.

	// +optional
	ReleaseName string `json:"releaseName,omitempty"`

	// Version is the version of the Helm chart. If it is not specified, the latest version will be used.
	// +optional
	Version string `json:"version,omitempty"`

	// Namespace is the namespace the Helm release will be installed on each selected Cluster. If it is not specified, the default namespace.
	// +optional
	Namespace string `json:"namespace,omitempty"`

	// Values is an inline YAML representing the values for the Helm chart. This YAML supports Go templating to reference
	// fields from each selected workload Cluster and programatically create and set values.
	// +optional
	Values string `json:"values,omitempty"`
}

// ClusterSelectorLabel defines a key/value pair used to select Clusters with a label matching the specified key and value.
type ClusterSelectorLabel struct {
	// Key is the label key.
	Key string `json:"key"`

	// Value is the label value.
	Value string `json:"value"`
}

// HelmChartProxyStatus defines the observed state of HelmChartProxy.
type HelmChartProxyStatus struct {
	// Ready is true when the HelmReleaseProxySpec for each selected Cluster is up to date.
	// +optional

> 	Ready bool `json:"ready"`

Reviewer 5: i wonder if we should use an enumeration of some sort here instead of a `bool`. although it might seem like we only need a ready/not-ready signal now, there might be cases in the future where we would need more states here.

Author: Right now I'm working on adding conditions to show more info about the status. For HelmChartProxy, I'm adding a `HelmReleaseProxySpecsUpToDateCondition` to ensure that the HelmChartProxy controller was able to create, update, and clean up HelmReleaseProxies. And I have a second condition `HelmReleaseProxiesReadyCondition` that aggregates the ready condition of HelmReleaseProxies. It's meant to be similar to how MachineSets have a `MachinesCreated` and a `MachinesReady` condition.

Reviewer 5: would those conditions be in addition to this bool or something different?

Reviewer 6: should align with other capi components and use readyCondition.

	ChartName string `json:"chartName,omitempty"`

	// RepoURL is the URL of the Helm chart repository.
	RepoURL string `json:"repoURL,omitempty"`

	// ReleaseName is the release name of the installed Helm chart. If it is not specified, a name will be generated.
	// +optional
	ReleaseName string `json:"releaseName,omitempty"`

	// Version is the version of the Helm chart. To be replaced with a compatibility matrix.
	// +optional
	Version string `json:"version,omitempty"`

	// Namespace is the namespace the Helm release will be installed on the referenced Cluster. If it is not specified, the default namespace.
	// +optional
	Namespace string `json:"namespace,omitempty"`

	// Values is an inline YAML representing the values for the Helm chart. This YAML is the result of the rendered
	// Go templating with the values from the referenced workload Cluster.
	// +optional
	Values string `json:"values,omitempty"`
}

// HelmReleaseProxyStatus defines the observed state of HelmReleaseProxy.
type HelmReleaseProxyStatus struct {
	// Ready is true when the Helm release on the referenced Cluster is up to date with the HelmReleaseProxySpec.
	// +optional

> 	Ready bool `json:"ready"`

Reviewer 5: similar question here about the `bool` usage, should we consider possible alternate states in the future?

Author: Similar to my comment above, I'm working on adding a `ClusterAvailableCondition` to signal that we can access the Cluster and get its kubeconfig and a `HelmReleaseReadyCondition` to show that the Helm release was successfully installed. Does that answer your question about adding additional states or do you think we should add something explicit like `Phase` to the status?

Reviewer 5: it does answer my question, i was hoping to see us not use a `bool` here, but it sounds like you've got an alternative idea about getting the deeper status from the conditions. i'm assuming those conditions would be on the `status` field then?

Author: I'll just answer both questions here since it's about the same thing in both CRDs. I've added a `clusterv1.Conditions` field below in the status structs like this.
```
// Conditions defines current state of the HelmReleaseProxy.
// +optional
Conditions clusterv1.Conditions `json:"conditions,omitempty"`
```

 I did have a question about the difference between the `Ready` boolean and the `ReadyCondition` which is aggregated using `conditions.SetSummary()`. From what I've seen most CAPI CRD statuses have a `Ready` field and a `FailureReason` and `FailureMessage` field, but it also includes a `ReadyCondition` which itself has its own `Reason` and `Message` fields. Is there a reason to have both, and if so, what's the difference in meaning and usage? If the goal is to deprecate the `Ready` field then maybe we can drop it from the new struct type here.

Reviewer 5: thanks for the explanation, it makes sense and is also confusing XD

i think the questions you are asking are great, i'm curious about those answers too. 

> If the goal is to deprecate the Ready field then maybe we can drop it from the new struct type here.

this seems important, because yeah maybe we can just avoid the whole field. it's not a blocker for me, i think it would be good if we drop the boolean but i think your reasoning for the conditions and the field make sense.

Reviewer 2: I *think* the bool fields are legacy and the conditions the future

(cc @fabriziopandini to be sure)

Author: Yep, if we're sure we want to replace the `ready` bool in favor of a summary condition then I can drop it from the status.

- Delete the Helm release if the HelmReleaseProxy is deleted.

This design of HelmReleaseProxy is especially convenient for the controller because a `Create()` maps directly to a `helm install` operation, a `Update()` maps directly to a `helm upgrade` operation, and a `Delete()` maps directly to a `helm uninstall` operation.

Note that the HelmReleaseProxy controller contains a static configuration and is only responsible for ensuring there is a Helm release matching the HelmReleaseProxy. The HelmReleaseProxy gets created, updated, and deleted by the HelmChartProxy controller.

#### Defining the list of Cluster add-ons to install

HelmChartProxy allows users to define the add-ons to install and which clusters they should be installed on using the `clusterSelector`. The specified Helm chart is then installed on every cluster that includes a label matching the cluster selector and makes it part of the orchestration. If a selected cluster no longer matches the cluster selector label, then the Helm release will be uninstalled. 

#### Installing and lifecycle managing the package manager

> ClusterAddonProvider for Helm imports Helm v3 as a library, making this a no-op. However, it is worth noting that each version of Helm only supports certain version of Kubernetes. For example, the [Helm docs](https://helm.sh/docs/topics/version_skew/#supported-version-skew) mention that Helm 3.9 supports Kubernetes 1.21 through 1.24. As a result, there may need to be multiple versions of HelmChartProxy and HelmReleaseProxy in order to be backwards compatible.

Reviewer 2: ```suggestion
ClusterAddonProvider for Helm imports Helm v3 as a library, making this a no-op. However, it is worth noting that each version of Helm only supports certain version of Kubernetes. For example, the [Helm docs](https://helm.sh/docs/topics/version_skew/#supported-version-skew) mention that Helm 3.9 supports Kubernetes 1.21 through 1.24.
```

Sorry, I meant dropping this sentence as well in my comment.

I think it would be confusing to have multiple different apiVersions of HelmChartProxy and HelmReleaseProxy to cover different Kubernetes versions. The sentence afterwards was fine

>  In this case, we will document the version support matrix for ClusterAddonProvider for Helm if it differs from the CAPI support matrix.

---
title: Cluster API Add-Ons Orchestration
authors:
  - "@fabriziopandini"
  - "@jont828"
  - "@jackfrancis"

> reviewers:

Reviewer: Make sure to add reviewers here as they review this & previous doc

- To make add-on management in Cluster API modular and pluggable, and to make it simple for developers to build a Cluster API Add-on Provider based on any package management tool, just like with infrastructure and bootstrap providers.

### Non goals

- To implement a full fledged package management tool in Cluster API; there are already several awesome package management tools in the ecosystem, and CAPI should not reinvent the wheel. 
- To provide a mechanism for altering, customizing, or dealing with single Kubernetes resources defining a Cluster add-on, i.e. Deployments, Services, ServiceAccounts. Cluster API should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool
- To expect users to use a specific package management tool.

### Future work

- Handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version.
- Provide support for advanced orchestration use cases leveraging on Cluster API’s lifecycle hooks for Cluster API.

> - Introduce capabilities to reconcile add-ons in an order that respects dependencies between add-ons.

Author: I think that's an issue I still have to consider. While Helm will install a package's critical dependencies automatically, I could see a situation where you'd want to wait for one Helm chart to be installed and ready before installing another, i.e. install some networking tool before installing another package that would need its functionality.

> - Introduce capabilities to reconcile add-ons in an order that respects dependencies between add-ons.

Reviewer 7: Maybe we should specify in the future work section what items refer only to the helm implementation and what do to the general `ClusterAddonProvider` concept (which will include the helm provider)?

- Deprecate or remove the ClusterResourceSet experimental feature.

## Proposal

This document introduces the concept of a ClusterAddonProvider, a new pluggable component to be deployed on the Management Cluster. A ClusterAddonProvider must be implemented using a specified package management tool and will act as an intermediary/proxy between Cluster API and the chosen package management solution.

> This proposal also introduces a ClusterAddonProvider for [Helm](https://helm.sh/) as a concrete example of how a ClusterAddonProvider would work with a package management tool. Users can configure ClusterAddonProvider for Helm through a new Kubernetes CRD called HelmChartProxy to be deployed on the Management Cluster. HelmChartProxy will act as an intermediary (proxy) between Cluster API and the Helm to install packages onto workload clusters. The implementation of ClusterAddonProvider for Helm can be used as a reference to implement ClusterAddonProviders for other package managers, i.e. ClusterAddonProvider for kapp.

Reviewer: That's a good question. There was actually a discussion in the original doc to allow installing addons in the management cluster as well, for example, cluster-autoscaler. Maybe this should be clarified in non-goals / future work if it's not going to be in scope initially.

> This proposal also introduces a ClusterAddonProvider for [Helm](https://helm.sh/) as a concrete example of how a ClusterAddonProvider would work with a package management tool. Users can configure ClusterAddonProvider for Helm through a new Kubernetes CRD called HelmChartProxy to be deployed on the Management Cluster. HelmChartProxy will act as an intermediary (proxy) between Cluster API and the Helm to install packages onto workload clusters. The implementation of ClusterAddonProvider for Helm can be used as a reference to implement ClusterAddonProviders for other package managers, i.e. ClusterAddonProvider for kapp.

Author: Yes, the current design of this project doesn't support installing add-ons on the management cluster.

#### Story 3

As a user/operator, I would like Cluster add-on providers to use information available in the Management Cluster (e.g. service or pod CIDR on the Cluster object) for customizing add-ons configurations specifically for the newly created workload Cluster.

#### Story 4

As a user/operator, I would like Cluster add-on providers to update add-ons configurations when information available in the Management Cluster (e.g add-on configuration) changes.

### Design and Responsibilities

Like the relationship between Cluster API and infrastructure providers, we can think of Cluster API Add-ons as a general problem with > 1 concrete implementations, each of which are pluggable components developed in repositories separated from Cluster API core.

However, considering the fact that the link between Cluster lifecycle and Cluster add-on lifecycle is simpler than e.g the link between a Cluster and its infrastructure provider, in this case it is possible to adopt a simplified interaction model between Cluster API and HelmChartProxy (vs an integration model based on CRD with a set of well known field defined in the CAPI contract).

> In the first iteration, we will implement this as a single, concrete HelmChartProxy CRD, with a controller watching Cluster API resources, mainly the Cluster and eventually the ControlPlane object.

Reviewer: +1, it would help to clarify by making a clear separate section for Helm Implementation

> In the first iteration, we will implement this as a single, concrete HelmChartProxy CRD, with a controller watching Cluster API resources, mainly the Cluster and eventually the ControlPlane object.

Reviewer 4: I agree we can do a better job in separating generic approach and specific implementation, but given the feedback we got while working on the google doc, we already know that readers can shave problems when reading the description of an abstract entity.

What i suggest is to move here the list of generic responsibilities currently the motivation, and build a short paragraph about the generic ClusterAddonProvider.

The current sentence and the following should be moved instead into ClusterAddonProvider for Helm

> In the first iteration, we will implement this as a single, concrete HelmChartProxy CRD, with a controller watching Cluster API resources, mainly the Cluster and eventually the ControlPlane object.

Reviewer 7: That's fair, and maybe reading the details of the helm implementation is a requisite to fully understand what a `ClusterAddonProvider` is.

That said, I don't think that should stop us from trying to define `ClusterAddonProvider` independently from a concrete implementation. Specially for this situation, where there is no programmatic contract (no API, etc.), I think we can really benefit from clearly defining `ClusterAddonProvider`'s responsibilities and constrains (if there are any). That's what other providers would need to design their own implementation.

This simple approach makes it possible to implement HelmChartProxy immediately and explore some foundational problems, however there will be a follow up iteration where HelmChartProxy will take advantage of Cluster API Runtime Extension and Lifecycle Hooks for supporting advanced orchestration use cases. 

Following paragraphs provide a deep dive on the responsibilities of the ClusterAddonProvider component; we are using Helm as a first target package management, but similar solutions can be defined for other tools as well. 

#### ClusterAddonProvider for Helm

Helm is a CNCF graduated project implementing a package manager widely used in the community; Helm packages are named charts.

ClusterAddonProvider for Helm could be implemented as a kubebuilder controller with a single reconciler watching for the Cluster object; each reconciliation will trigger Helm operations by leveraging on  https://github.com/helm/helm imported as a library.

> The ClusterAddonProvider for Helm is going to implement two new CRDs (see next paragraphs for the full spec). The first is named HelmChartProxy used to specify a Helm chart with configuration values and a selector identifying the clusters the add-on applies to. The second is called HelmReleaseProxy and contains information about the underlying Helm release installed on a cluster from a Helm chart. All the configurations will be done from the HelmChartProxy and the HelmReleaseProxy is simply used to maintain an inventory as well as surface information.

Author: Yes, it's a CRD created and deleted by the controller. From a UX standpoint, HelmChartProxies are the primary "interface" for users to configure their Helm charts and HelmReleaseProxies are a way to introspect the status of a given Helm release on the workload cluster without having to get the kubeconfig and go into the Cluster.

#### Defining the list of Cluster add-ons to be installed

ClusterAddonProvider for Helm will use HelmChartProxy to define the addons to install and which clusters they should be installed on. Each HelmChartProxy will contain fields to specify a single Helm chart. Additionally, the spec will contain a key value pair which will be used as a cluster selector The Helm chart is then installed on every cluster that includes a label matching the cluster selector and makes it part of the orchestration. If a selected cluster no longer matches the cluster selector label, then the Helm release will be uninstalled. 

Note that the cluster selector label differs from the match label field in ClusterResourceSets as HelmChartProxy specifies a single key value pair whereas the match labels field matches any number of labels. The match labels field was not used because by adding another label pair, all of the currently matched clusters would be unselected and that would trigger the deletion of all associated Helm releases. 

> This solution can be generalized to other ClusterAddonProviders by defining a CRD that specifies an add-on to be installed and includes a cluster selector. The controller would opt-in any Cluster matching the selector for orchestration: install the add-on if it is not present, update it if it is present and configuration in the CRD changed, and uninstall the add-on if the Cluster is no longer selected.

Author: That's a good question. My intention is to design a cluster selector that providers can use easily, but in theory they could implement one however they like so long as it has a way to select and unselect a Cluster. Part of the reason I designed it with one label instead of the MatchLabels functionality though, is that by adding additional labels to MatchLabels, you'd unintentionally unselect every Cluster that previously matched.

> This solution can be generalized to other ClusterAddonProviders by defining a CRD that specifies an add-on to be installed and includes a cluster selector. The controller would opt-in any Cluster matching the selector for orchestration: install the add-on if it is not present, update it if it is present and configuration in the CRD changed, and uninstall the add-on if the Cluster is no longer selected.

Reviewer 4: +1 to drop the sentence.
Other providers can use the same approach or implement their own strategy to defining the list of Cluster add-ons to be installed

> This solution can be generalized to other ClusterAddonProviders by defining a CRD that specifies an add-on to be installed and includes a cluster selector. The controller would opt-in any Cluster matching the selector for orchestration: install the add-on if it is not present, update it if it is present and configuration in the CRD changed, and uninstall the add-on if the Cluster is no longer selected.

Reviewer 2: +1

#### Install and lifecycle manage the Package Manager

Some Package Manager might require in-cluster components, like e.g. the kapp-controller, however this is not the case for Helm starting from v3, so this is a no-op.

> In case this is required, the ClusterAddonProvider should take care of this task as well.

Author: Sure thing. The way ClusterAddonProvider for Helm works is that it imports Helm v3 as a library in Go. Based on this design, if we wanted to upgrade to a future v4 of Helm, it would require releasing a new version of ClusterAddonProvider for Helm.

> In case this is required, the ClusterAddonProvider should take care of this task as well.

Reviewer 7: Right, that makes sense. But I guess I was more talking about the generic `ClusterAddonProvider`. We are adding the tools installation (when necessary) to its responsibilities. Should we add upgrades as well?

 
On top of that, in many Cluster API installations, the add-on version an operator would like to use is limited to a list of certified versions defined in each organization/product vendor.

We can rely upon existing functionality in Helm (“Capabilities”) to enforce Kubernetes version-specific requirements in a chart, including whether or not to install anything at all.

#### Generating package configurations 

Each package manager has its own specific configuration format, like e.g. Helm’s values.yaml file; on top of that, usually each add-on package has its own set of specific keys/data structures defining configuration values, like e.g. aws-calico configuration.

For ClusterAddonProvider for Helm, a solution is relatively simple. The Helm CLI allows a chart to be configured using a key/value map passed in as CLI arguments, i.e. `--set “key1=value1,key2=value2`. As a result, we can include a similar key/value map in HelmChartProxy for configuring package values.

> HelmChartProxy can use an approach similar to templated variables in ClusterClass to fill in variables values. The key/value map can include Go templating representing fields from the Cluster definition, i.e. cluster name, namespace, control plane ref, pod CIDRs. These template fields will be resolved dynamically based on the actual values for each workload cluster selected by HelmChartProxy’s cluster selector.

Reviewer 7: Is this how we are proposing that all ClusterAddonProvider's solve this problem? We might not need to prescriptive but I think it's important to avoid re-inventing the wheel. 

With a proposed common solution to a general problem we can do things like provide packages with tools to implement it. This should make implementing new ClusterAddonProvider easier.

Author: Not necessarily, I can imagine a situation where other package managers don't configure their values the same way as Helm. The only part I'm prescriptive of is that we'd want to be able to configure the package by referencing fields from the selected Cluster using Go templating.

Reviewer 4: I agree there is value in creating a common solution to a general problem, but at the same time, it seems to me a little bit out of the scope of this proposal, that in some ways, it is exploring this field at this stage.

I think that follow-up iterations of this work should determine what is common/reusable and refactor things accordingly

The solution implemented in the ClusterAddonProvider for Helm is to use HelmReleaseProxy to maintain an inventory (see prototype design for more specifics). Each HelmReleaseProxy corresponds to a Helm release installed through HelmChartProxy. The controller will only orchestrate Helm releases with a corresponding HelmReleaseProxy, ensuring that out of band Helm releases won’t be affected. 

Additionally, each HelmReleaseProxy will include a label indicating the Cluster the release is installed on, and an additional label indicating the HelmChartProxy it corresponds to. This provides a way to query for the Helm releases installed on a Cluster and also the Helm releases installed from a HelmChartProxy. This implementation also generalizes to any ClusterAddonProvider as it uses a CRD to maintain the inventory instead of the package manager.

#### Surface Cluster Add-ons Status

> The add-on orchestrator is responsible for installing add-ons on an arbitrary number of workload clusters. As such, it would be beneficial to see the status of an add-on from the management cluster and avoid having to fetch the kubeconfig of the workload cluster.

Reviewer 7: is "add-on orchestrator" the same as ClusterAddonProvider?

Author: Yep, I can reword that

Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.

## Alternatives

### Why not ClusterResourceSets?

This proposal introduces ClusterAddonProviders as opposed to overhauling ClusterResourceSets for the following reasons:

- ClusterResourceSets do not follow the Cluster API convention of using a declarative spec and having controllers create, delete, or update resources to build the desired state. Even though there has been work to add a Reconcile or an "apply always" mode to ClusterResourceSets, it is [explicitly stated](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20200220-cluster-resource-set.md#non-goalsfuture-work) that deleting resources from Clusters is a non-goal.
- Users are already using package management tools, and being able to use their preferred package management tool reduces user friction towards adopting an add-on solution.
- ClusterResourceSets have relatively fewer user experts compared to existing package management tools like Helm.

> - ClusterResourceSets fail to take advantage of functionatlity in existing package management tools like dependency installation, and evolving ClusterResourceSets would require re-inventing the functionality of existing package management tools.

Reviewer: ```suggestion
- ClusterResourceSets fail to take advantage of functionality in existing package management tools like dependency installation, and evolving ClusterResourceSets would require re-inventing the functionality of existing package management tools.
```

### Why not CRS

> Here are reasons why ClusterResourceSets are not a good option for cluster-api-idiomatic addons:

Reviewer 7: I think we are missing some of the most important ones:
* In order to improve the add-on management capabilities of CRS, we would need to re-invent the wheel, doing what a few other tools in the space are already doing.
* Users are already using add-ons management tools and they want to use the same ones for cluster add-ons.
* CRS design presents severe limitations around deleting resources and dynamic versioning based on the kubernetes version.

not blocking: some of the reasons listed here are a bit sketchy, like `require a ton of effort to do` or `require significant refactor work that may impact existing CRS users`. One could argue that implementing a `ClusterAddonProvider` is as complex or more than such refactor. No need to remove them (I 💯 agree ClusterAddonProvider is a better solution to the problem), but from a "why didn't we do this" debate perspective, some of these sound a bit forced.

- ClusterResourceSets do not orchestrate deletion of add-ons and was not designed with that in mind. In order to orchestrate deletion of add-ons, it would require a significant redesign of the code and interface. Doing so would still not address the above issues compared to designing a new solution with all of this in mind.

### Existing package management controllers

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer: > This info about the team later in the process of design

missing word maybe? not sure I understand the phrasing

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer 4: yep, what about:

This info reached the team later in the design process, ...

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer: my two cents: I would omit that part, as not knowing something early in the design process shouldn't be used in justification for not choosing one option over another (sunk cost fallacy), that's what the design process is for: finding the best possible implementation and iterating on it. However, if there are design considerations why one option is better to address the specific user stories that this work is focused on, that's what the proposal should surface.

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer 4: That's fair, +1 for me to to change the suggested middle sentence into

> Given that this proposal aims to explore the topic of orchestration between Cluster API and add-on tools, we still believe that using a simpler, homegrown version of HelmRelease CRD and avoiding dependency with other projects allows us to iterate faster while we are still figuring out the best orchestration patterns between the two components.

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer 8: Would it make sense to take a hybrid approach? come up with a custom api allowing experimentation, but map it to say, flux's helm release object in an initial prototype. Can then explore the space quickly without committing to much. Can then find out how well it works in the field, and if the api has its own gotchas beyond just being a simple wrapper. If it needs to be its own thing, then spin it out. If it ends up being basically a 1:1 mapping, deprecate it and keep shipping that component of flux?

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer 4: my first reaction to this idea is that we should close this iteration as it is, and take the proper time to evaluate this option for the next iteration (which is similar to what I had in mind while writing about gathering info to evaluate next steps).
but let's hear other opinions / the opinion of the team working on this

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Author: @fabriziopandini Just pushed some edits to this section based on what you wrote!

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer 9: > my two cents: I would omit that part, as not knowing something early in the design process shouldn't be used in justification for not choosing one option over another (sunk cost fallacy),

I agree. When I brought up the issue and started pointing these alternatives out I was hoping that we could use this information in the design of a cluster-api specific controller. I'm still open to helping the team evaluate further.

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Reviewer 4: @faiq thanks for your offer, personally I will really appreciate if we can have a demo/walk through of helm flux integration with CAPI in the near future, so we can brainstorm a little bit around what's next

> It is worth noting that there are other projects have integrated a package management tool with a Kubernetes controller. For example, Flux has developed a [Helm controller](https://github.com/fluxcd/helm-controller) as part of its GitOps toolkit for syncing Clusters with sources of configuration. It defines a `HelmRelease` CRD and offers [integration with Cluster API](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) where users can create a `HelmRelease` to install an add-on on a specified Cluster. While this proposal aims to develop a Cluster API native solution for add-ons, users can also explore well-established add-on solutions in the ecosystem for using add-ons with Cluster API.

Author: @fabriziopandini @faiq That sounds good to me. I'll resolve this thread and we can start a thread in the Slack if you want to brainstorm ideas of using Flux's Helm controller with CAPI as well.

#### Story 4

As a developer or Cluster operator, I would like Cluster add-on providers to use information available in the Management Cluster (e.g. service or pod CIDR on the Cluster object) for customizing add-ons configurations specifically for the newly created workload Cluster.

#### Story 5

As a developer or Cluster operator, I would like Cluster add-on providers to update add-ons configurations when information available in the Management Cluster (e.g add-on configuration) changes.

### ClusterAddonProvider Functionality

A ClusterAddonProvider is a new component for bringing add-ons into the Cluster API lifecycle. It serves as a general solution to Cluster API Add-ons with multiple concrete implementations, similar to the relationship between Cluster API and infrastructure providers. It aims to do so handle add-on lifecycle management in a holistic way by providing the following capabilities:

- Provide or install a package management tool
- Determine which add-ons to install or uninstall on each Cluster

> - Provide a default configuration for each add-on

Author: I think that was part of @fabriziopandini's original requirements but I believe the intention was to ensure that if we supply a blank `valuesTemplate`, we will install a Helm chart using the default values.yaml instead of setting them all to empty fields.

> - Provide a default configuration for each add-on

Reviewer 2: Makes sense. As far as I'm aware with Helm that's just how it works per default. Even if you provide an empty values.yaml it would just use the default from the Helm Chart.

> - Provide a default configuration for each add-on

Reviewer 4: imagine a user that wants to customize default values in a helm chart, this goal is about making this simple by allowing to do the customization one instead of for each cluster.

> - Provide a default configuration for each add-on

Reviewer 2: Ah now I've got it :). So essentially the valuesTemplate in the HelmChartProxy in the current implementation?

> - Provide a default configuration for each add-on

Author: Yep pretty much.

> - Provide a default configuration for each add-on

Reviewer 2: Alright. Fine to resolve for me

  // +optional
  MatchingClusters []corev1.ObjectReference `json:"matchingClusters"`
}
```

#### HelmReleaseProxy

HelmReleaseProxy represent a single Helm release installed on a selected workload Cluster. HelmReleaseProxies are created and managed by the HelmChartProxy controller. The parent HelmChartProxy will be responsible for keeping its HelmReleaseProxy children up-to-date as its lifecycle is updated (for example, if the `values` properties of a HelmChartProxy are updated, its child HelmReleaseProxy resources will be updated as well).

```go
// HelmReleaseProxySpec defines the desired state of HelmReleaseProxy.
type HelmReleaseProxySpec struct {
  // ClusterRef is a reference to the Cluster to install the Helm release on.
  ClusterRef *corev1.ObjectReference `json:"clusterRef"`

  // ChartName is the name of the Helm chart in the repository.
  ChartName string `json:"chartName"`

  // RepoURL is the URL of the Helm chart repository.
  RepoURL string `json:"repoURL"`

  // ReleaseName is the release name of the installed Helm chart. If it is not specified, a
  // name will be generated.
  // +optional

>   ReleaseName string `json:"releaseName,omitempty"`

Author: The name is generated in the Helm install client and after we generate it, we go back and set it in the spec. I'm not super concerned about the implementation details right now, but the idea is that if it's not set when initialized, it'll be set to a generated value.

>   ReleaseName string `json:"releaseName,omitempty"`

Reviewer 2: Alright that answers my main question that it's tracked here. Definitely fine to figure out details later. I could imagine it's easier to just generate the ReleaseName in the default webhook of this CRD and then use it in the helm command. This way we don't have to reconcile back into the spec. But as you said that's an implementation detail.

  // Cluster. If it is not specified, the default namespace will be used.
  // +optional
  ReleaseNamespace string `json:"namespace,omitempty"`

  // Version is the version of the Helm chart. If it is not specified, the chart will use 
  // and be kept up to date with the latest version.
  // +optional
  Version string `json:"version,omitempty"`

  // Values is an inline YAML representing the values for the Helm chart. This YAML is the result of the rendered
  // Go templating with the values from the referenced workload Cluster.
  // +optional
  Values string `json:"values,omitempty"`
}

// HelmReleaseProxyStatus defines the observed state of HelmReleaseProxy.
type HelmReleaseProxyStatus struct {
  // Conditions defines current state of the HelmReleaseProxy.
  // +optional
  Conditions clusterv1.Conditions `json:"conditions,omitempty"`

  // Status is the current status of the Helm release.
  // +optional
  Status string `json:"status,omitempty"`

  // Revision is the current revision of the Helm release.
  // +optional
  Revision int `json:"revision,omitempty"`

  // Namespace is the namespace of the Helm release on the workload cluster.
  // +optional

>   Namespace string `json:"namespace,omitempty"`

Author: The namespace in the spec corresponds to the namespace argument in `helm install --namespace`, which is described as the "namespace scope for this request". The namespace in the status corresponds to the output of `helm status` shown below. I'm not sure if it's guaranteed to be identical but I just based it off of the Helm API.

```
NAME: nginx-ingress-1663364363
LAST DEPLOYED: Fri Sep 16 17:39:26 2022
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
The NGINX Ingress Controller has been installed.
```

>   Namespace string `json:"namespace,omitempty"`

Reviewer 2: I would have assumed that it's always the same. Essentially if I want to deploy a chart in a certain namespace I would expect it to end up there :)

>   Namespace string `json:"namespace,omitempty"`

Author: Fair enough, I can ask on the Helm repo to clarify. I figured they probably had a reason to put it in the status (in case there's some special edge cases).

>   Namespace string `json:"namespace,omitempty"`

Reviewer 2: Would be interesting to know

>   Namespace string `json:"namespace,omitempty"`

Author: I opened an issue on the Helm repo to clarify but I think we can probably assume it's the same. In that case do you want me to just drop the field from the status and reference the field in the spec?

>   Namespace string `json:"namespace,omitempty"`

Reviewer 2: I think it makes sense to drop it for now. We can re-add it later if we find out that it is important.

What do you mean with:
> reference the field in the spec

>   Namespace string `json:"namespace,omitempty"`

Author: Similar to the release name, if the field isn't set, it'll be set to a default or generated value by the webhook/controller.

>   Namespace string `json:"namespace,omitempty"`

Reviewer 2: I think it would be good to have a authoritative value in the spec (similar to the release name).

I would either make it mandatory or default it to the same value as the helm cli (which is probably "default")

>   Namespace string `json:"namespace,omitempty"`

Author: Fair enough, I'll make it required and we can always change it later if there's an additional edge case we didn't consider.

>   Namespace string `json:"namespace,omitempty"`

Reviewer 2: One nit. As of today kubectl does client-side OpenAPI validation per default (it can be disabled with --validate=false). The client-side validation will be dropped in the future.

But as of today making a field mandatory in the API type will force most of our users (who don't know about --validate=false or don't want to use it) to set the field on client-side.

So as of today if you want to make it optional for the regular kubectl user you have to keep it optional.

Going forward, I'm not sure how it works but probably the mutating webhooks are called before OpenAPI validation (no guarantees).

Just wanted to provide context, up to you and it's v1alpha1 so we can change it over time.

It probably makes sense though to define both ReleaseName and RleeaseNamespace the same way

>   Namespace string `json:"namespace,omitempty"`

Author: Actually I'll probably keep the ReleaseNamespace optional and have a defaulting webhook set it in the HelmReleaseProxy if we need to. But yeah I'm open to iterating on how we want to handle fields like this.

>   Namespace string `json:"namespace,omitempty"`

Reviewer 2: Sounds good to me

In future iterations, we plan to handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version by leveraging the new runtime hooks feature.

#### Managing dependencies

Managing dependencies between HelmChartProxies is out of scope for this project. However, Helm charts define their dependencies in the [Chart.yaml file](https://helm.sh/docs/topics/charts/#the-chartyaml-file) and will install them automatically. If a Helm chart has a critical dependency on another chart, it is the chart owner's responsibility to enable it by default.

### Example: Installing Calico CNI

Developers and Cluster operators often want to install a CNI on their workload clusters. HelmChartProxy controller can be used to install the Calico CNI on each selected workload cluster. In this example, the value configuration can be leveraged to initialize the `installation.calicoNetwork.ipPools` array by creating an element for every pod CIDR in the Cluster.

```yaml

> apiVersion: addons.cluster.x-k8s.io/v1beta1

Author: Gotcha. I chose v1beta1 so it could align with the versions of other CAPI resources. Would there be any problems from having different apiVersions?

> apiVersion: addons.cluster.x-k8s.io/v1beta1

Reviewer 2: It's totally fine to have a different apiVersion.

- Introduce capabilities to reconcile add-ons in an order that respects dependencies between add-ons.
- Deprecate or remove the ClusterResourceSet experimental feature.

## Proposal

This document introduces the concept of a ClusterAddonProvider, a new pluggable component to be deployed on the Management Cluster. A ClusterAddonProvider must be implemented using a specified package management tool and will act as an intermediary/proxy between Cluster API and the chosen package management solution.

This proposal also introduces a ClusterAddonProvider for [Helm](https://helm.sh/) as a concrete example of how a ClusterAddonProvider would work with a package management tool. Users can configure ClusterAddonProvider for Helm through a new Kubernetes CRD called HelmChartProxy to be deployed on the Management Cluster. HelmChartProxy will act as an intermediary (proxy) between Cluster API and the Helm to install packages onto workload clusters. The implementation of ClusterAddonProvider for Helm can be used as a reference to implement ClusterAddonProviders for other package managers, i.e. ClusterAddonProvider for kapp.

### User Stories

#### Story 1

> As a user/operator, I would like Cluster add-on providers to automatically include or install the package management tool.

Reviewer 6: user is an ambiguous term, It'd be clear to refer to specific personas e.g cluster operator, dev...

#### Story 4

As a user/operator, I would like Cluster add-on providers to use information available in the Management Cluster (e.g. service or pod CIDR on the Cluster object) for customizing add-ons configurations specifically for the newly created workload Cluster.

#### Story 5

As a user/operator, I would like Cluster add-on providers to update add-ons configurations when information available in the Management Cluster (e.g add-on configuration) changes.

### ClusterAddonProvider Functionality

A ClusterAddonProvider is a new component for bringing add-ons into the Cluster API lifecycle. It serves as a general solution to Cluster API Add-ons with multiple concrete implementations, similar to the relationship between Cluster API and infrastructure providers. It aims to do so handle add-on lifecycle management in a holistic way by providing the following capabilities:

- Provide or install a package management tool
- Determine which add-ons to install on each Cluster
- Provide a default configuration for each add-on
- Configure an add-on by resolving or inherit information from the Cluster configuration (i.e. service or pod CIDR for CNI)

> - Install add-ons during cluster creation (immediately after the API server is installed). This operation usually brings in some additional requirements such as: 

Reviewer 6: >- Install add-ons during cluster creation (immediately after the API server is installed)

how is this ensured?

Author: It's not currently ensured in this first iteration. Maybe @fabriziopandini can comment on what he had in mind?

Reviewer 4: I think that if we phrase this as (immediately after the API server is available) this will be more clear.
WRT to implementation, we can look at control-plane or cluster conditions for now; in a future iteration we will rely on lifecycle hooks

### Implementation details 

The implementation will include two CRDs and two controllers as follows. As this is a prototype, the specific details regarding the CRD specs are subject to change.

#### HelmChartProxy CRD

The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

```go
// HelmChartProxySpec defines the desired state of HelmChartProxy.
type HelmChartProxySpec struct {
	// ClusterSelector selects Clusters with a label that matches the specified key/value pair. The Helm chart will be
	// installed on all selected Clusters. If a Cluster is no longer selected, the Helm release will be uninstalled.
	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`

	// ChartName is the name of the Helm chart in the repository.

> 	ChartName string `json:"chartName,omitempty"`

Reviewer 6: nit: I assume these are required, so remove `omitempty`

> 	ChartName string `json:"chartName,omitempty"`

Reviewer 6: Does chartName adds value over RepoURL?
Also should it be just Name aligning with RepoURL and Version?

Author: Helm requires a chartName along with a repoURL since a repo can have multiple charts.

	// +optional
	Version string `json:"version,omitempty"`

	// Namespace is the namespace the Helm release will be installed on each selected Cluster. If it is not specified, the default namespace.
	// +optional
	Namespace string `json:"namespace,omitempty"`

	// Values is an inline YAML representing the values for the Helm chart. This YAML supports Go templating to reference
	// fields from each selected workload Cluster and programatically create and set values.
	// +optional
	Values string `json:"values,omitempty"`
}

// ClusterSelectorLabel defines a key/value pair used to select Clusters with a label matching the specified key and value.
type ClusterSelectorLabel struct {
	// Key is the label key.
	Key string `json:"key"`

	// Value is the label value.
	Value string `json:"value"`
}

// HelmChartProxyStatus defines the observed state of HelmChartProxy.
type HelmChartProxyStatus struct {
	// Ready is true when the HelmReleaseProxySpec for each selected Cluster is up to date.
	// +optional
	Ready bool `json:"ready"`

	// MatchingClusters is the list of references to Clusters selected by the ClusterSelectorLabel.
	// +optional

> 	MatchingClusters []corev1.ObjectReference `json:"matchingClusters"`

Reviewer 6: may be just .Clusters?
what's the value of this field?

Author: The idea is that you can verify that you didn't make a mistake by installing on a Cluster you didn't intend to or forgot to include a Cluster. Since we're selecting using labels to select, it could be easy to forget to remove a label in a Cluster for example.

	// Conditions defines current state of the HelmChartProxy.
	// +optional
	Conditions clusterv1.Conditions `json:"conditions,omitempty"`

	// FailureReason will be set in the event that there is a an error reconciling the HelmChartProxy.
	// +optional

> 	FailureReason string `json:"failureReason,omitempty"`

Reviewer 6: why do you need this over conditions?

Author: This was paired with the `ready` bool, I think we can probably just drop them both in that case

}
```

#### HelmReleaseProxy CRD

The second CRD, HelmReleaseProxy, is responsible for installing and orchestrating the Helm release on a selected workload Cluster. The HelmReleaseProxy is created as part of the parent HelmChartProxy flow. The parent HelmChartProxy will be responsible for keeping its HelmReleaseProxy children up-to-date as its lifecycle is updated (for example, if the `values` properties of a HelmChartProxy are updated, its child HelmReleaseProxy resources will be updated as well).

```go
// HelmReleaseProxySpec defines the desired state of HelmReleaseProxy.
type HelmReleaseProxySpec struct {
	// ClusterRef is a reference to the Cluster to install the Helm release on.

> 	ClusterRef *corev1.ObjectReference `json:"clusterRef,omitempty"`

Reviewer 6: is this cross namespace reference or should this be a local / local typed ref?

Reviewer 2: (somewhat related to my point above: are those CRDs namespaced or cluster-wide)

Author: Currently there's nothing to restrict it to the same namespace. Would it be beneficial to restrict it to the same namespace?

Reviewer 4: I'm +1 to have this namespaced for starting. It is simpler and we are consistent with everything else in CAPI.

Author: Sounds good. TBH I'm a bit on the fence about it but I agree that we can move forward with making it namespaced for simplicity and consistency with CAPI.

#### Considerations about the CRD design

The scope of the cluster values accessible using Go templating must be defined along with use cases. An option would be to expose the Builtin struct type used in ClusterClass or construct a similar type. Additionally, while resources like the control plane and cluster have a 1:1 correlation with each selected cluster, resources like MachineDeployments or Machines could have any number and can’t be selected in a similar way. As a result, additional work is needed to decide if there’s a use case for including information in those types.

### Example: Installing Calico CNI

Developers and Cluster operators often want to install a CNI on their workload clusters. HelmChartProxy controller can be used to install the Calico CNI on each selected workload cluster. In this example, the value configuration can be leveraged to initialize the `installation.calicoNetwork.ipPools` array by creating an element for every pod CIDR in the Cluster.

```yaml
apiVersion: addons.cluster.x-k8s.io/v1beta1
kind: HelmChartProxy
metadata:

>   name: calico-cni

Reviewer 6: In which namespace should I deploy my chart, where and how does the watcher controller run?

Author: Right now you can deploy HelmChartProxy to any namespace, and the controller is run in a pod on the management Cluster similar to a CAPI provider

spec:

>   clusterSelector:

Reviewer 6: does this apply to all namespaces?

Author: At the moment it does, though I've been thinking about adding a `namespaces` or `allowedNamespaces` field to the selector.

  // ReleaseNamespace is the namespace the Helm release will be installed on each selected
  // Cluster. If it is not specified, the default namespace.
  // +optional
  ReleaseNamespace string `json:"namespace,omitempty"`

  // Version is the version of the Helm chart. If it is not specified, the chart will use 
  // and be kept up to date with the latest version.
  // +optional
  Version string `json:"version,omitempty"`

  // ValuesTemplate is an inline YAML representing the values for the Helm chart. This YAML supports Go templating to reference
  // fields from each selected workload Cluster and programatically create and set values.
  // +optional
  ValuesTemplate string `json:"valuesTemplate,omitempty"`
}

// HelmChartProxyStatus defines the observed state of HelmChartProxy.
type HelmChartProxyStatus struct {
  // Conditions defines current state of the HelmChartProxy.
  // +optional
  Conditions clusterv1.Conditions `json:"conditions,omitempty"`

  // MatchingClusters is the list of references to Clusters selected by the ClusterSelector.
  // +optional
  MatchingClusters []corev1.ObjectReference `json:"matchingClusters"`
}

> ```

Reviewer 6: does this have a status?

Author: Do you mean the `MatchingClusters` field or the HelmChartProxyStatus?

- Provide or install a package management tool
- Determine which add-ons to install or uninstall on each Cluster
- Provide a default configuration for each add-on
- Configure an add-on by resolving or inherit information from the Cluster configuration (i.e. service or pod CIDR for CNI)
- Install add-ons during cluster creation (immediately after the API server is installed). This operation usually brings in some additional requirements such as: 
  - Resolve the add-on version for the Cluster’s Kubernetes version
  - Wait for the add-on to be actually running
- Provide a mechanism for upgrading an add-on before, during, or after the Cluster upgrade and cleaning up the previous version
- Support for out-of band upgrades for add-ons not linked to the Cluster lifecycle, i.e. CVE fix
- Configure pre and post installation steps for each add-on, thus allowing to manage migrations/operations such as:
  - CPI migration from in-tree to out-of-tree
  - CoreDNS configuration migration from one version to another

> - Orchestrate add-on deletion before Cluster deletion

Author: I was discussing with @CecileRobertMichon about this point. Is there a use case for why we'd want to delete the add-on if the Cluster is getting delete anyway? Or in other words would we ever expect users to `helm delete` their charts before deleting a Cluster?

Reviewer 2: I think as long as they are stateless or their data is pushed/pulled to somewhere else anyway we shouldn't care. I think this applies to 99% of the cases we currently consider (including stuff like log shipping and metrics collection).

Author: Gotcha, I guess my question then is do we want to explicitly set an owner ref from HelmReleaseProxy or instead reword this point a bit? If we do set an ownerRef, then that forces it to be in the same namespace as the Cluster.

Reviewer 2: We should probably decide first if the CRDs are namespaced or cluster-wide. This seems to be a major point and it has influence on a lot of the currently open discussions.

Reviewer 10: I think adding a namespace match in addition to the cluster label match perhaps makes things overly complicated. This is also a feature we can add later if folks decide they want to compose cluster label key/value + namespace.

So my vote would be to omit the cluster namespace and default to "cluster-wide".

ClusterAddonProvider for Helm proposes two new CRDs named HelmChartProxy and HelmReleaseProxy. HelmChartProxy is used to specify a Helm chart with configuration values and a selector identifying the Clusters the add-on applies to. HelmReleaseProxy contains information about the underlying Helm release installed on a cluster from a Helm chart. All the configurations will be done from the HelmChartProxy and the HelmReleaseProxy is simply used to maintain an inventory as well as surface information. Note that the specific fields in both CRDs are subject to change as the project evolves.

This is a link to the prototype for [ClusterAddonProvider for Helm](https://github.com/Jont828/cluster-api-addon-provider-helm), and a demo can be found in this [Cluster API Office Hours recording](https://www.youtube.com/watch?v=dipR1Tzh4jE). The prototype implements HelmChartProxy and HelmReleaseProxy as well as controllers for both CRDs.

#### HelmChartProxy

HelmChartProxy serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

```go
// HelmChartProxySpec defines the desired state of HelmChartProxy.

> type HelmChartProxySpec struct {

Reviewer 9: Looking at this API type I see a lot of similarities to the [HelmReleases type implemented by Flux](https://fluxcd.io/flux/components/helm/helmreleases/)

To me, that raises a lot of questions:

- Are we reimplementing a solution the community has already come up with?
- How is a user who is already using flux to manage apps supposed to use this in conjunction with Flux?
- If we implemented a ClusterAddonProvider for flux and that would manage helm charts what would that look like?

I'm wondering if AddonProvider should be made even more generic? 

Thank you,

Author: Thanks for sharing your concerns. I saw the Flux Helm controller a while back though I'm not super familiar with how it works. To answer your question, the goal of this project is to bring an add-ons user story specifically to Cluster API, not to try to provide a generalized solution. More specifically, we'd like to be able to manage add-ons in CAPI by applying specs to the management Cluster and not having to manually enter the workload cluster and run `helm install`. 

Currently ClusterAddonProvider for Helm does not touch out-of-band Helm releases, meaning if a Helm release was installed from another source besides our controller, it won't be reconciled. So if you already installed Helm charts with Flux this shouldn't touch it. I'd have to look into the Flux controller more to say how we could create ClusterAddonProvider for Flux.

Reviewer 9: > To answer your question, the goal of this project is to bring an add-ons user story specifically to Cluster API, not to try to provide a generalized solution. More specifically, we'd like to be able to manage add-ons in CAPI by applying specs to the management Cluster and not having to manually enter the workload cluster and run helm install.

That's a great goal, but it sounds like this solution is somewhere in between cluster resource sets and full solution for applications that something like Flux or any other GitOps framework provides. In my experience, these are the types of solutions people end up with for application management, even in the case of applying applications to other clusters. 

Has there been any thought into delegating application deployment responsibilities to a solution like this entirely?

Reviewer 9: For example, there's a whole project that is supported, tested, and maintained by the flux project that handles just the Helm bits https://github.com/fluxcd/helm-operator

I'm worried that the `ClusterAddonProvider` controllers will run into the same issues these projects have already encountered.

Author: I don't have too much context about the flux project's work with Helm. Could you share a bit about the issues the those projects already encountered? Tagging @fabriziopandini as well.

Reviewer 8: if the flux helm-operator could be tweaked with a minor change to deploy stuff in the target cluster, that seems like a nice approach. Its already a related cncf project.

Reviewer 9: > I don't have too much context about the flux project's work with Helm. Could you share a bit about the issues the those projects already encountered? Tagging @fabriziopandini as well.

To be honest, any bug fix that they've implemented can count as an issue they've encountered. I'm saying this because this implementation is essentially doing what the helm-operator has already accomplished and tested in production clusters. 

Tools like this should be considered in the Alternatives section. An effort should be made to see if we can easily integrate an already existing and well tested component, rather than writing a new one. 

I'm not sure of the timeline to further evaluate tools here, but I can help out flesh out these sections.

Reviewer 9: Flux has several examples of propogating helm-releases to remote clusters including one for [cluster-api that are clearly laid out in the documentation](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api). There's also examples of using [karmada in a multi-cluster environment as well](https://fluxcd.io/flux/use-cases/karmada/). 

At the very least, this needs to be documented in the alternatives section or brought together as a different proposal that can create a way for existing solutions or homegrown ones such as the `HelmProxy`to work with cluster API.

Reviewer 8: > Flux has several examples of propogating helm-releases to remote clusters including one for [cluster-api that are clearly laid out in the documentation](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api). There's also examples of using [karmada in a multi-cluster environment as well](https://fluxcd.io/flux/use-cases/karmada/).

Thats a very interesting example, thanks. Bundling the flux helm controller along with cluster-api seems like maybe an easy/viable solution to the problem. It worked well to bundle cert-manager to reduce project load.

Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.

## Alternatives

### Why not ClusterResourceSets?

This proposal introduces ClusterAddonProviders as opposed to overhauling ClusterResourceSets for the following reasons:

- ClusterResourceSets violate the Cluster API design philosophy of defining a desired state through a declarative spec and having controllers create, delete, or update resources to build the desired state. Instead, ClusterrResourceSets are designed to "apply once and forget."
- Users are already using package management tools, and being able to use their preferred package management tool reduces user friction towards adopting an add-on solution.
- ClusterResourceSets have relatively fewer user experts compared to existing package management tools like Helm.
- ClusterResourceSets fail to take advantage of functionatlity in existing package management tools like dependency installation, and evolving ClusterResourceSets would require re-inventing the functionality of existing package management tools.

> - ClusterResourceSets do not orchestrate deletion of add-ons and was not designed with that in mind. In order to orchestrate deletion of add-ons, it would require a significant redesign of the code and interface. Doing so would still not address the above issues compared to designing a new solution with all of this in mind.

Reviewer 9: ```suggestion
- ClusterResourceSets do not orchestrate deletion of add-ons and was not designed with that in mind. In order to orchestrate deletion of add-ons, it would require a significant redesign of the code and interface. Doing so would still not address the above issues compared to designing a new solution with all of this in mind.

### Flux

- Another alternative is to package Flux with cluster-api. Flux is a tool for keeping Kubernetes clusters in sync with sources of configuration (like Git repositories), and automating updates to configuration when there is new code to deploy.
- Flux offers [direct integration with cluster-API ](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api) by using a field built in `HelmRelease` API type. In its current implementation, it would require a `HelmRelease` per application per cluster.
```

- To design a solution for orchestrating Cluster add-ons.
- To leverage existing package management tools such as Helm for all the foundational capabilities of add-on management, i.e. add-on packages/repository, templating/configuration, add-on creation, upgrade and deletion etc. 
- To make add-on management in Cluster API modular and pluggable, and to make it simple for developers to build a Cluster API Add-on Provider based on any package management tool, just like with infrastructure and bootstrap providers.

### Non goals

- To implement a full fledged package management tool in Cluster API; there are already several awesome package management tools in the ecosystem, and CAPI should not reinvent the wheel. 
- To provide a mechanism for altering, customizing, or dealing with single Kubernetes resources defining a Cluster add-on, i.e. Deployments, Services, ServiceAccounts. Cluster API should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool.
- To expect users to use a specific package management tool.
- To implement a solution for installing add-ons on the management cluster itself.

### Future work

> - Handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version.

Reviewer 4: @Jont828  am I wrong or this is already covered by the current implementation

Author: We don't really have a way to trigger a Helm upgrade when a Cluster is upgraded. You can set a new chart version in HelmChartProxy and it can update on the next reconciliation though.

Note that the cluster selector label differs from the match label field in ClusterResourceSets as HelmChartProxy specifies a single key value pair whereas the match labels field matches any number of labels. The match labels field was not used because by adding another label pair, all of the currently matched clusters would be unselected and that would trigger the deletion of all associated Helm releases. 

This solution can be generalized to other ClusterAddonProviders by defining a CRD that specifies an add-on to be installed and includes a cluster selector. The controller would opt-in any Cluster matching the selector for orchestration: install the add-on if it is not present, update it if it is present and configuration in the CRD changed, and uninstall the add-on if the Cluster is no longer selected.

#### Install and lifecycle manage the Package Manager

Some Package Manager might require in-cluster components, like e.g. the kapp-controller, however this is not the case for Helm starting from v3, so this is a no-op.
In case this is required, the ClusterAddonProvider should take care of this task as well.

> In the case of Helm, we will need to consider that each version of Helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.

Author: The HelmChartProxy controller imports Helm v3 as a library. From what I saw in [these docs](https://helm.sh/docs/topics/version_skew/#:~:text=As%20of%20Helm%203%2C%20Helm,n%2D1%20versions%20of%20Kubernetes), Helm 3 has limitations with which K8s versions it supports, but we can't really control that. So we'd need to pick the right version of Helm to use for an older K8s version.

> In the case of Helm, we will need to consider that each version of Helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.

Reviewer 2: Okay and how does this related to "versions of HelmChartProxy"?

> In the case of Helm, we will need to consider that each version of Helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.

Author: I'm not exactly sure, I think was part of the original draft @fabriziopandini wrote, maybe he can comment about what he envisioned?

> In the case of Helm, we will need to consider that each version of Helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.

Reviewer 4: If I remember well, this is just a reminder to document the support matrix of this new component if this does not overlap with CAPI support matrix

> In the case of Helm, we will need to consider that each version of Helm has official support for 4 versions of Kubernetes. Because a HelmChartProxy will require a particular implementation of the helm (and downstream Kubernetes) libraries, we will want to expose that version support matrix when we maintain versions of HelmChartProxy itself.

Reviewer 2: Yup definitely something to keep in mind. On the other hand it might just be what they officially support (and test?) and it would also work with older/newer versions as long as they don't require new features. (somewhat like client-go also kind of works as long as you just use stable stuff)

Mapping Kubernetes version to add-ons version
Cluster add-ons by definition have a lifecycle strictly linked to the Cluster lifecycle, and the main consequences of this is the fact that there is a strict correlation between the Cluster’s Kubernetes version and the add-on version to be installed in a Cluster.

In some cases this correlation is explicit, like e.g. for CPI following the “Versioning Policy for External Cloud Providers”; in other cases the compatibility matrix is defined case by case in documentation, like for “Supported version” in the Calico documentation.
 
On top of that, in many Cluster API installations, the add-on version an operator would like to use is limited to a list of certified versions defined in each organization/product vendor.

We can rely upon existing functionality and capabilities in Helm to enforce Kubernetes version-specific requirements in a chart, including whether or not to install anything at all. It's worth noting that other add-on tools might not provide this functionality out of the box.

#### Generating package configurations 

> Each package manager has its own specific configuration format, like e.g. Helm’s values.yaml file; on top of that, usually each add-on package has its own set of specific keys/data structures defining configuration values, like e.g. aws-calico configuration.

Author: I think it's just a Helm chart since I found [this link](https://docs.aws.amazon.com/eks/latest/userguide/calico.html) when Googling it. But I believe the idea is just that the values.yaml file can have any number of nested fields.

> Each package manager has its own specific configuration format, like e.g. Helm’s values.yaml file; on top of that, usually each add-on package has its own set of specific keys/data structures defining configuration values, like e.g. aws-calico configuration.

Reviewer 2: Ah makes sense. The example (without link) might be a bit confusing

> Each package manager has its own specific configuration format, like e.g. Helm’s values.yaml file; on top of that, usually each add-on package has its own set of specific keys/data structures defining configuration values, like e.g. aws-calico configuration.

Reviewer: +1 on adding a link

> For ClusterAddonProvider for Helm, a solution is relatively simple. The Helm CLI allows a chart to be configured using a key/value map passed in as CLI arguments, i.e. `--set “key1=value1,key2=value2`. As a result, we can include a similar key/value map in HelmChartProxy for configuring package values.

Author: Yep, I made some changes recently to the value config so I'll update this section as well.

Additionally, each HelmReleaseProxy will include a label indicating the Cluster the release is installed on, and an additional label indicating the HelmChartProxy it corresponds to. This provides a way to query for the Helm releases installed on a Cluster and also the Helm releases installed from a HelmChartProxy. This implementation also generalizes to any ClusterAddonProvider as it uses a CRD to maintain the inventory instead of the package manager.

#### Surface Cluster Add-ons Status

ClusterAddonProviders are responsible for installing add-ons on an arbitrary number of workload clusters. As such, it would be beneficial to see the status of an add-on from the management cluster and avoid having to fetch the kubeconfig of the workload cluster.

In the ClusterAddonProvider for Helm, the status of a Helm release can be conveyed through the status field of a HelmReleaseProxy. It includes a ready status, which conveys that the HelmRelease is installed and up to date with the CRD spec and a failure reason which is set if there is an error in reconciliation.

> To convey additional information, the HelmReleaseProxy status can include an array of `clusterv1.Conditions` and define new condition types as needed to offer more specific information. For example, there could be a `PodsReady` and `DeploymentsReady` condition.

Author: Yeah when I wrote that the former was what I had implemented/committed too and the latter was more speculative. In that example, there's a PR in Helm for conveying the status of the deployment and pods in their library and if that PR merges then we can add additional conditions to convey that.

#### Upgrade strategy

> Users will be able to upgrade the add-ons across multiple Clusters at once since updating the version in the HelmChartProxy triggers a Helm upgrade for all the Helm releases it manages. Additionally, HelmChartProxy allows users to upgrade their add-ons in a declarative manner.

Author: Yeah I'm aware that the current upgrade story isn't ideal. I'm planning on revisiting this in a future iteration to flesh out a more comprehensive solution, but for the first iteration I wanted to point out how the current functionality can help.

Note that hooks and automatic triggers for add-on upgrades when a Cluster is upgraded to a new Kubernetes version is out of scope for this iteration. Users will need to determine if an add-on version upgrade is needed, and if so, set a compatible version in the HelmChartProxy.

In future iterations, we plan to handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version by leveraging the new runtime hooks feature.

#### Managing dependencies

Managing dependencies between HelmChartProxies is out of scope for this project. However, Helm charts define their dependencies in the [Chart.yaml file](https://helm.sh/docs/topics/charts/#the-chartyaml-file) and will install them automatically. If a Helm chart has a critical dependency on another chart, it is the chart owner's responsibility to enable it by default.

### Implementation details 

The implementation will include two CRDs and two controllers as follows. As this is a prototype, the specific details regarding the CRD specs are subject to change.

#### HelmChartProxy CRD

> The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

Author: I don't believe it's namespaced currently. HelmChartProxy should be able to select a Cluster across namespaces and do a Helm install since all it really needs is to fetch the spec and the kubeconfig and then initialize a Helm client. What would be the pros and cons of making it namespaced?

Tagging @elmiko since it's related to the namespace issue of the `clusterRef` in HelmReleaseProxy.

> The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

Reviewer 2: cc @fabriziopandini

> The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

Reviewer 2: Hm good question, not sure about the trade-offs. Currently all our CRDs except ExtensionConfig are namespaced.

Could have impact on `cluster move / backup / restore` (if it's in scope of that).

Maybe also relevant regarding RBAC if you want to give folks access to only one cluster (+ all its resouces) + it's addon CRs, that's probably impossible with a cluster-wide CRD.

On the other side of course it would limit the usefulness as the CRDs would probably only work with the Clusters in a single namespace.

Not sure if we have some historic background that could help here.

> The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

Author: One idea I'm fleshing out is to add a  `namespaces` field to the `clusterSelector` based on [this](https://capz.sigs.k8s.io/topics/multitenancy.html#allowednamespaces). The idea would be that you can

1. Default to the namespace of the HCP if it's not set
2. Pass in a list of namespaces to search for Clusters
3. Use something like a wildcard `*` to search across all namespaces 

WDYT about something along those lines?

> The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

Reviewer 2: It highly depends on the CRDs being namespaced or cluster-wide. If they are namespaced, the CRs should probably only act on the CAPI CRDs in the same namespace (like other CAPI CRDs)

> The first CRD, HelmChartProxy, serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 

Author: Gotcha. I'd like to move forward with the CRDs being cluster-wide for ease of use (so you don't need 100 HelmChartProxies if you have 100 namespaces), but I'm open to making them namespaced if there's a strong case for doing so.

```go
// HelmChartProxySpec defines the desired state of HelmChartProxy.
type HelmChartProxySpec struct {
	// ClusterSelector selects Clusters with a label that matches the specified key/value pair. The Helm chart will be
	// installed on all selected Clusters. If a Cluster is no longer selected, the Helm release will be uninstalled.

> 	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`

Author: The reason being is that the `metav1.LabelSelector` works with an "and" on a set of labels instead of an "or." So if my label selector `{key1=value1}` and my clusters have `key1=value1` and then I set it to `{key1=value1, key2=value2}` then all my clusters become unselected. From a UX perspective I think that's a really awkward case where adding something triggers deletion.

> 	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`

Reviewer 2: It's the same awkward UX then every other LabelSelector in Kubernetes. I personally don't have a problem with that as it's consistent with every label selector. Looking at what we can currently achieve with only a single label supporting a regular LabelSelector would make the selection more powerful.

> 	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`

Author: When I was working on the label selection, I couldn't really find any examples of projects that use a label selector to orchestrate creation AND deletion, the use cases I could find were searching for Machines like in CAPI or with CRS where we only apply to what matches (so being "unmatched" doesn't really pose a UX problem). I decided to use a single label to simplify it based on how [AAD Pod Identity](https://docs.microsoft.com/en-us/azure/aks/use-azure-ad-pod-identity#run-a-sample-application) uses a single label to create AADPodIdentityBindings. 

Do you know of any projects that orchestrate deletion using the `metav1.LabelSelector`?

> 	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`

Reviewer 2: No, but I'm not sure if that's relevant.

You can also change your single label selector and it would trigger deletion. So I don't see how restriction functionality improves on this.

> 	ClusterSelector ClusterSelectorLabel `json:"clusterSelector"`

Author: Thanks for sharing those examples. In that case I'll change it so we can go with a metav1.LabelSelector for selection.

	// ChartName is the name of the Helm chart in the repository.
	ChartName string `json:"chartName,omitempty"`

	// RepoURL is the URL of the Helm chart repository.
	RepoURL string `json:"repoURL,omitempty"`

	// ReleaseName is the release name of the installed Helm chart. If it is not specified, a name will be generated.
	// +optional

> 	ReleaseName string `json:"releaseName,omitempty"`

Author: It's the name of a release on a Helm level. If you set this field, all the Helm releases installed through this HemChartProxy will have just have the same release name, otherwise it'll be a auto-generated name

	// Version is the version of the Helm chart. If it is not specified, the latest version will be used.
	// +optional

> 	Version string `json:"version,omitempty"`

Author: Yep in my implementation we're checking to see if there's a "diff" in the versions.

> 	Version string `json:"version,omitempty"`

Author: If you don't specify a version Helm will pull the latest chart version and you can see that version in the metadata so all we have to do is compare it to the existing install.

> 	Version string `json:"version,omitempty"`

Reviewer 2: Sounds perfect. Should we possibly extend that comment slightly to make that 100% clear?

	// Namespace is the namespace the Helm release will be installed on each selected Cluster. If it is not specified, the default namespace.
	// +optional
	Namespace string `json:"namespace,omitempty"`

	// Values is an inline YAML representing the values for the Helm chart. This YAML supports Go templating to reference
	// fields from each selected workload Cluster and programatically create and set values.
	// +optional

> 	Values string `json:"values,omitempty"`

Author: Yeah the `--set` field isn't used anymore, I just forgot to remove references to it in the doc.

}

// HelmReleaseProxyStatus defines the observed state of HelmReleaseProxy.
type HelmReleaseProxyStatus struct {
	// Ready is true when the Helm release on the referenced Cluster is up to date with the HelmReleaseProxySpec.
	// +optional
	Ready bool `json:"ready"`

	// FailureReason will be set in the event that there is a an error reconciling the HelmReleaseProxy.
	// +optional
	FailureReason string `json:"failureReason,omitempty"`

	// Conditions defines current state of the HelmReleaseProxy.
	// +optional
	Conditions clusterv1.Conditions `json:"conditions,omitempty"`

	// Status is the current status of the Helm release.
	// +optional
	Status string `json:"status,omitempty"`

	// Revision is the current revision of the Helm release.
	// +optional
	Revision int `json:"revision,omitempty"`

	// Namespace is the namespace of the Helm release on the workload cluster.
	// +optional
	Namespace string `json:"namespace,omitempty"`
}
```

#### Controller Design

The HelmChartProxy controller is responsible for maintaining our list of HelmReleaseProxies and resolving any referenced Cluster fields. The controller will:

- Find all workload clusters matching the cluster selector.

> - Resolve the referenced Cluster fields in the values map based on each selected cluster.

Reviewer 2: nit/Q Slightly confused by "map" we are running go templating here, right?

- Install a Helm release on the workload cluster if it does not exist.
- Update the existing Helm release if the HelmReleaseProxy values or version do not match the existing Helm release.
- Delete the Helm release if the HelmReleaseProxy is deleted.

Note that the HelmReleaseProxy controller contains a static configuration and is only responsible for ensuring there is a Helm release matching the HelmReleaseProxy. The HelmReleaseProxy gets created, updated, and deleted by the HelmChartProxy controller.

#### Considerations about the CRD design

The scope of the cluster values accessible using Go templating must be defined along with use cases. An option would be to expose the Builtin struct type used in ClusterClass or construct a similar type. Additionally, while resources like the control plane and cluster have a 1:1 correlation with each selected cluster, resources like MachineDeployments or Machines could have any number and can’t be selected in a similar way. As a result, additional work is needed to decide if there’s a use case for including information in those types.

> It would also be useful to add `clusterv1.Conditions` to the HelmChartProxy and HelmReleaseProxy status fields to surface additional information about the status of the Helm release in a way that is consistent with the design of existing Cluster API CRDs.

Reviewer 2: nit: seems slightly out of date, given that the structs above have Conditions

  chartName: tigera-operator
  values: |
    installation:
      cni:
        type: Calico
        ipam:
          type: HostLocal
      calicoNetwork:
        bgp: Disabled
        mtu: 1350
        ipPools:{{range $i, $cidr := .Cluster.Spec.ClusterNetwork.Pods.CIDRBlocks }}
        - cidr: {{ $cidr }}
          encapsulation: None
          natOutgoing: Enabled
          nodeSelector: all(){{end}}
```

## Cluster API enhancements

Despite the fact that most of the heavy lifting for add-on management is delegated to external ClusterAddonProvider components, over time we would like to provide Cluster API’s users an integrated and seamless user experience for add-on orchestration.

### clusterctl

ClusterAddonProvider is a new component to be installed in a Management Cluster, and in order to make this operation simpler, we can extend clusterctl in order to manage ClusterAddonProvider implementations in clusterctl init, clusterctl upgrade, etc.

Note: clusterctl assumes all the providers to be linked to a CAPI contract; while it is not clear at this stage if the same applies to ClusterAddonProvider. This should be further investigated.

> 

Reviewer 2: Q: Are there any security implications / considerations that are worth mentioning?

Essentially we are introducing a component like tiller from helm v2 which was dropped because of security issues. I guess with the goals of an addon controller we basically have to run it in "god mode" (regarding the workload clusters) and it's probably fine as not everyone has access to the management cluster.

Anyway, just curious if there are any considerations worth mentioning (e.g. maybe something like auditing could be useful to provide visibility into what was done by the addon provider  in the future)

Author: I'm not too familiar with security in K8s. I believe @fabriziopandini added this section so I'll tag him. In terms of "god mode" I'm not too clear on why/if the add-on controller would be more insecure than the CRS controller if it the idea is to select clusters to install some arbitrary YAML.

Reviewer 4: Not sure if grasp how the security concern relates to the paragraph above on clusterctl, but I'm ok in documenting that this  component is going to act on the workload clusters using the admin kubeconfig existing in the management cluster (but TBH this is the same of what any other component in CAPI is doing).

Reviewer 2: I just added the comment here as the security section we have in other proposals not present. If we think there's nothing worth mentioning feel free to resolve.
